syncapi/types/types.go

```diff
@@ -223,151 +231,121 @@ func (t *TopologyToken) Decrement() {
+		PDUPosition: pduPos,
 	}
 }
+
 func NewTopologyTokenFromString(tok string) (token TopologyToken, err error) {
-	t, _, err := newSyncTokenFromString(tok)
-	if err != nil {
+	if len(tok) < 1 {
+		err = fmt.Errorf("empty topology token")
 		return
 	}
-	if t.Type != SyncTokenTypeTopology {
-		err = fmt.Errorf("token %s is not a topology token", tok)
+	if tok[0] != SyncTokenTypeTopology[0] {
+		err = fmt.Errorf("topology token must start with 't'")
 		return
 	}
-	if len(t.Positions) < 2 {
-		err = fmt.Errorf("token %s wrong number of values, got %d want at least 2", tok, len(t.Positions))
-		return
+	parts := strings.Split(tok[1:], "_")
+	var positions [2]StreamPosition
+	for i, p := range parts {
+		if i > len(positions) {
+			break
+		}
+		var pos int
+		pos, err = strconv.Atoi(p)
+		if err != nil {
+			return
+		}
+		positions[i] = StreamPosition(pos)
 	}
-	return TopologyToken{
-		syncToken: *t,
-	}, nil
+	token = TopologyToken{
+		Depth:       positions[0],
+		PDUPosition: positions[1],
+	}
+	return
 }
 
 // NewStreamToken creates a new sync token for /sync
-func NewStreamToken(pduPos, eduPos StreamPosition, logs map[string]*LogPosition) StreamingToken {
+func NewStreamToken(
+	pduPos, typingPos, receiptPos, sendToDevicePos StreamPosition,
+	logs map[string]*LogPosition,
+) StreamingToken {
 	if logs == nil {
 		logs = make(map[string]*LogPosition)
 	}
 	return StreamingToken{
-		syncToken: syncToken{
-			Type:      SyncTokenTypeStream,
-			Positions: []StreamPosition{pduPos, eduPos},
-		},
-		logs: logs,
+		PDUPosition:          pduPos,
+		TypingPosition:       typingPos,
+		ReceiptPosition:      receiptPos,
+		SendToDevicePosition: sendToDevicePos,
+		logs:                 logs,
 	}
 }
+
 func NewStreamTokenFromString(tok string) (token StreamingToken, err error) {
-	t, categories, err := newSyncTokenFromString(tok)
-	if err != nil {
+	if len(tok) < 1 {
+		err = fmt.Errorf("empty stream token")
 		return
 	}
-	if t.Type != SyncTokenTypeStream {
-		err = fmt.Errorf("token %s is not a streaming token", tok)
+	if tok[0] != SyncTokenTypeStream[0] {
+		err = fmt.Errorf("stream token must start with 's'")
 		return
 	}
-	if len(t.Positions) < 2 {
-		err = fmt.Errorf("token %s wrong number of values, got %d want at least 2", tok, len(t.Positions))
-		return
-	}
-	logs := make(map[string]*LogPosition)
-	if len(categories) > 1 {
-		// dl-0-1234
-		// $log_name-$partition-$offset
-		for _, logStr := range categories[1:] {
-			segments := strings.Split(logStr, "-")
-			if len(segments) != 3 {
-				err = fmt.Errorf("token %s - invalid log: %s", tok, logStr)
-				return
-			}
-			var partition int64
-			partition, err = strconv.ParseInt(segments[1], 10, 32)
-			if err != nil {
-				return
-			}
-			var offset int64
-			offset, err = strconv.ParseInt(segments[2], 10, 64)
-			if err != nil {
-				return
-			}
-			logs[segments[0]] = &LogPosition{
-				Partition: int32(partition),
-				Offset:    offset,
-			}
+	categories := strings.Split(tok[1:], ".")
```

Comment:
This will panic if `len(tok) == 1`?

Reply:
It shouldn't do, as it will just result in a zero-length string slice. The tests seem OK with it.